File: solutions.digamma.damas.jcr/src/main/java/solutions/digamma/damas/jcr/model/JcrPathFinder.java
package solutions.digamma.damas.jcr.model;

import solutions.digamma.damas.common.WorkspaceException;
import solutions.digamma.damas.entity.Entity;
import solutions.digamma.damas.common.MisuseException;
import solutions.digamma.damas.content.PathFinder;
import solutions.digamma.damas.auth.Token;
import solutions.digamma.damas.inspection.NotNull;
import solutions.digamma.damas.jcr.common.Exceptions;
import solutions.digamma.damas.jcr.session.SessionUser;
import solutions.digamma.damas.jcr.session.SessionWrapper;

import javax.jcr.RepositoryException;
import javax.jcr.Session;
import java.nio.file.Paths;

/**
 * @author Ahmad Shahwan
 */
public interface JcrPathFinder<T extends Entity>
        extends SessionUser, PathFinder<T> {

    @Override
    default T find(@NotNull Token token, @NotNull  String path)
            throws WorkspaceException {
        try (SessionWrapper session = openSession(token)) {
            path = Paths.get(path).normalize().toString();
            if (path.startsWith("../")) {
                throw new MisuseException("Invalid relative path.");
            }
            path = path.equals("") ? "." : path;
            return this.find(session.getSession(), path);
        } catch (RepositoryException e) {
            throw Exceptions.convert(e);
        }
    }

    /**
     * Perform path look-up.
     *
     * @param session
     * @param path
     * @return
     * @throws RepositoryException
     * @throws WorkspaceException
     */
    T find(@NotNull Session session, String path)
            throws RepositoryException, WorkspaceException;
}

File: solutions.digamma.damas.jcr/src/main/kotlin/solutions/digamma/damas/jcr/model/JcrPathFinder.kt
package solutions.digamma.damas.jcr.model

import solutions.digamma.damas.common.MisuseException
import solutions.digamma.damas.common.WorkspaceException
import solutions.digamma.damas.content.PathFinder
import solutions.digamma.damas.entity.Entity
import solutions.digamma.damas.jcr.common.Exceptions
import solutions.digamma.damas.jcr.session.JcrSessionConsumer
import java.nio.file.Paths
import javax.jcr.RepositoryException

/**
 * An interface that provide path finding functionality.
 *
 * @author Ahmad Shahwan
 */
internal interface JcrPathFinder<T : Entity> : JcrSessionConsumer, PathFinder<T> {

    /**
     * This method converts a thrown exception into a [WorkspaceException].
     * It also normalizes the path to a JCR compatible one.
     */
    @Throws(WorkspaceException::class)
    override fun find(path: String): T = Exceptions.check {
        var p = Paths.get(path).normalize().toString()
        if (p.startsWith("../")) {
            throw MisuseException("Invalid relative path.")
        }
        if (p == "") p = "."
        this.doFind(p)
    }

    /**
     * Perform path look-up.
     *
     * @param path
     * @return
     * @throws RepositoryException
     * @throws WorkspaceException
     */
    @Throws(RepositoryException::class, WorkspaceException::class)
    fun doFind(path: String): T
}
